br/pkg/storage/gcs_extra.go

```diff
@@ -0,0 +1,428 @@
+	UploadId string   `xml:"UploadId"`
+}
+
+type Part struct {
+	Text       string `xml:",chardata"`
+	PartNumber int    `xml:"PartNumber"`
+	ETag       string `xml:"ETag"`
+}
+
+type CompleteMultipartUpload struct {
+	XMLName xml.Name `xml:"CompleteMultipartUpload"`
+	Text    string   `xml:",chardata"`
+	Parts   []Part   `xml:"Part"`
+}
+
+func (w *GCSWriter) finalizeXMLMPU() error {
+	finalXMLRoot := CompleteMultipartUpload{
+		Parts: []Part{},
+	}
+	slices.SortFunc(w.xmlMPUParts, func(a, b *xmlMPUPart) int {
+		return a.partNumber - b.partNumber
+	})
+	for _, part := range w.xmlMPUParts {
+		part := Part{
+			PartNumber: part.partNumber,
+			ETag:       part.etag,
+		}
+		finalXMLRoot.Parts = append(finalXMLRoot.Parts, part)
+	}
+
+	xmlBytes, err := xml.Marshal(finalXMLRoot)
+	if err != nil {
+		return fmt.Errorf("failed to encode XML: %v", err)
+	}
+
+	opts := &storage.SignedURLOptions{
+		Scheme:          storage.SigningSchemeV4,
+		Method:          "POST",
+		Expires:         time.Now().Add(w.signedURLExpiry),
+		QueryParameters: url.Values{mpuUploadIDQuery: []string{w.uploadID}},
+	}
+	u, err := w.cli.Bucket(w.bucket).SignedURL(w.blob, opts)
+	if err != nil {
+		return fmt.Errorf("Bucket(%q).SignedURL: %s", w.bucket, err)
+	}
+
+	client := resty.New()
+	resp, err := client.R().SetBody(xmlBytes).Post(u)
+	if err != nil {
+		return fmt.Errorf("POST request failed: %s", err)
+	}
+
+	if resp.StatusCode() != http.StatusOK {
+		return fmt.Errorf("POST request returned non-OK status: %d", resp.StatusCode())
+	}
+	return nil
+}
+
+type chunk struct {
+	buf     []byte
+	num     int
+	cleanup func()
+}
+
+func (w *GCSWriter) appendMPUPart(part *xmlMPUPart) {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	w.xmlMPUParts = append(w.xmlMPUParts, part)
+}
+
+func (w *GCSWriter) cancel() error {
+	opts := &storage.SignedURLOptions{
+		Scheme:          storage.SigningSchemeV4,
+		Method:          "DELETE",
+		Expires:         time.Now().Add(w.signedURLExpiry),
+		QueryParameters: url.Values{mpuUploadIDQuery: []string{w.uploadID}},
+	}
+	u, err := w.cli.Bucket(w.bucket).SignedURL(w.blob, opts)
+	if err != nil {
+		return fmt.Errorf("Bucket(%q).SignedURL: %s", w.bucket, err)
+	}
+
+	client := resty.New()
+	resp, err := client.R().Delete(u)
+	if err != nil {
+		return fmt.Errorf("POST request failed: %s", err)
+	}
+
+	if resp.StatusCode() != http.StatusOK {
+		return fmt.Errorf("POST request returned non-OK status: %d", resp.StatusCode())
+	}
+
+	return nil
+}
+
+type xmlMPUPart struct {
+	uploadBase
+	buf        []byte
+	uploadID   string
+	partNumber int
+	checksum   string
+	etag       string
+	finished   bool
+}
+
+func (p *xmlMPUPart) Clone() *xmlMPUPart {
+	return &xmlMPUPart{
+		uploadBase: p.uploadBase,
+		uploadID:   p.uploadID,
+		buf:        p.buf,
+		partNumber: p.partNumber,
+		checksum:   p.checksum,
+	}
+}
+
+func (p *xmlMPUPart) Upload() error {
+	err := p.upload()
+	if err == nil {
+		return nil
+	}
+
+	for i := 0; i < p.retry; i++ {
```

Comment:
use backoffer?

Reply:
for above and this comments: these codes are copied from https://github.com/liqiuqing/gcsmpu